### PR TITLE
MSC4033: Explicit ordering of events for receipts

### LINKED ISSUE #1167: Clarify if notifications are cleared by stream ordering or topological ordering
**Link to problem area**: https://spec.matrix.org/latest/client-server-api/#receiving-notifications

**Issue**:

It is not stated clearly which notifications get cleared when a read receipt is received, currently the spec reads:

> When the user updates their read receipt (either by using the API or by sending an event), notifications prior to and including that event MUST be marked as read.

It is unclear if "prior to" refers to stream ordering or topological ordering. Synapse clears this based on stream ordering.

(Note that because of the inconsistency in `/sync` and `/messages` either could give odd results to a user.)

See conversation around https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$sb6NPRyuiu4wN7egyY8SJ6YoubvQqZvJSOlMFsTzVSE?via=matrix.org&via=libera.chat&via=element.io

### PR DESCRIPTION
**Pull request by andybalaam**
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/andybalaam/event-thread-and-order/proposals/4033-event-thread-and-order.md)

Points out ambiguities in the current spec about what a receipt means, because it is not clear on the order of events that means they are "before" or "after" a receipt. Proposes that events and receipts contain an `order` property that makes it explicit which receipt marks which event as read.

Fixes https://github.com/matrix-org/matrix-spec/issues/1167

### CODE REVIEW

File: proposals/4033-event-thread-and-order.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+is read.
+
+### Proposed definition of *after*
+
+We propose that the definition of *after* should be:
+
+* Event A is after event B if its *Stream Order* is larger.
+
+We define *Stream Order* to be an immutable, unique number attached to an event
+on creation that defines the order of events in regard to receipts.
```
Comment (Johennes) [inline proposals/4033-event-thread-and-order.md:122]: Where exactly does this happen between clients and home servers? Is this "on creation" on the receiver's home server?

File: proposals/4033-event-thread-and-order.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+Because it controls the meaning of read receipts, it is desirable that Stream
+Order be as close as possible to Sync Order, the order in which clients receive
+events via sync. However, since clients receive events in different orders
+depending on the APIs they use, it is not a goal that Stream Order exactly match
+the order in which clients receive events. Instead, it provides a canonical
+order that means we can be clear about what the user has read, and thus should
+generally increase for "newer" messages. Clients may decide to re-order events
+into Stream Order, or they may decide to display unread messages higher up the
+timeline if the orders do not match the order they choose for display.
```
Comment (Johennes) [inline proposals/4033-event-thread-and-order.md:156]: Would clients not always want to display messages in the same order that their unread status is maintained in? Otherwise, their read receipts could mark messages that haven't actually been read as read. I'm also struggling slightly to imagine a UI treatment where read and unread messages are mingled.

File: proposals/4033-event-thread-and-order.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+thread branching from them. Non-thread children of thread roots (e.g.
+reactions to a thread root) are also in the `main` thread. This is a change to
+the current definition.
+
+### Proposed change in consideration of redacted events
+
+We propose that redacted events should never be considered unread.
+
+This avoids the need to identify which thread a redacted event belongs to, which
+will be difficult if its `thread_id` property has been stripped out.
```
Comment (Johennes) [inline proposals/4033-event-thread-and-order.md:192]: IIUC we could also mandate `thread_id` to not be stripped during redaction but this would require a new room version. If that's the case, it might be worth calling this out explicitly here.

File: proposals/4033-event-thread-and-order.md [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+We propose to modify the definitions slightly, and include extra information in
+the events provided by the server to make it easy to calculate whether an event
+is read.
+
+### Proposed definition of *after*
+
+We propose that the definition of *after* should be:
+
+* Event A is after event B if its *Stream Order* is larger.
```
Comment (turt2live) [inline proposals/4033-event-thread-and-order.md:119]: Note that at least in Synapse it's possible to get negative stream ordering values if the server backfilled the event. Stream ordering itself is not send to clients, so they might not be aware of the "source" for an event.

Comment (richvdh) [reply proposals/4033-event-thread-and-order.md:119]: (synapse uses -ve stream orderings for such events mostly to ensure they are never sent down `/sync`, fwiw)

File: proposals/4033-event-thread-and-order.md [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+[^1]: We think this is similar to Stream Ordering, which is mentioned once in
+  the spec without definition in
+  [7.6 Syncing](https://spec.matrix.org/unstable/client-server-api/#syncing),
+  but we are not certain that it is identical, because we believe it may be
+  possible for different clients to receive events in a different order from
+  each other for the same account. For example, if one client is doing an
+  incremental sync, and another is doing an initial sync, recently-arrived
+  events that are "old" in Topological Order may be received in different orders
+  on the two clients.
```
Comment (richvdh) [inline proposals/4033-event-thread-and-order.md:62]: I think "Sync ordering" is exactly the same as "Stream ordering" -- at least for Synapse. Synapse will always send events down `/sync` response in stream ordering. Different clients should never see events in `/sync` responses in different orders.

File: proposals/4033-event-thread-and-order.md [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+The current definition of *after* is ambiguous, and difficult for clients to
+calculate. It depends on only receiving events via sync, which is impractical
+due to backpagination, and the desire to fetch thread messages via the
+`relations` API.
+
+Further, we believe that the current working definition of "Sync Order" does not
+make sense, because different clients may receive events in different orders,
+meaning that a receipt from one client would be interpreted differently by
+another client.
```
Comment (richvdh) [inline proposals/4033-event-thread-and-order.md:97]: as above: I don't *think* this is possible.

File: proposals/4033-event-thread-and-order.md (hunk 2 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:156]: How clients choose to display late-arriving old messages (e.g. after a network split) is considered out of scope of this MSC. Some clients might want to show them at the bottom of the timeline to make sure the user can see them, and some clients might want to mix them in higher up the timeline, perhaps with some kind of bold marking to show that they are unread.

The point in this MSC is that the client can't do that unless it can determine whether a specific message is unread. This means that the order needs to be explicit, and actually that it can't be topological order, since that would make newly-arriving old messages look like they had been read when they have not.

File: proposals/4033-event-thread-and-order.md (hunk 3 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:192]: Yes correct. I will add something.

File: proposals/4033-event-thread-and-order.md [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+* it has a `thread_id` property
+
+Otherwise, it is in the `main` thread.
+
+No events are in more than one thread.
+
+Note: this means that thread roots are in the `main` thread, and not in the
+thread branching from them. Non-thread children of thread roots (e.g.
+reactions to a thread root) are also in the `main` thread. This is a change to
+the current definition.
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:185]: Yes, will add a sentence.

File: proposals/4033-event-thread-and-order.md (hunk 4 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:119]: I will add a note that stream order can be negative. I am arguing that stream ordering should be exposed to clients explicitly.

File: proposals/4033-event-thread-and-order.md (hunk 5 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:62]: OK. I am not making that assumption here for all home servers.

File: proposals/4033-event-thread-and-order.md (hunk 6 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:97]: OK. I don't know whether this is important enough to investigate or whether to simply remove this comment from the MSC. I don't this it is critical to justifying why we need explicit order.

**Comment by bradtgmurray** [comment]
Funnily enough we at Beeper have been using this exact approach (`com.beeper.hs.order` field in the `unsigned` portion) in our Synapse fork and I can say it's been useful for simplifying client development: https://github.com/beeper/synapse/pull/42

File: proposals/4033-event-thread-and-order.md [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+We propose that the definition of whether an event is read should be unchanged,
+but this note should be added:
+
+> (Because the receipt itself contains the `order` of the pointed-to event,
+> there is no need to examine the pointed-to event, so it is sufficient to
+> compare the `order` of the event in question with the `order` in the
+> receipt.)
+
+And the above definition of *after* should be included.
+
```
Comment (clokep) [inline proposals/4033-event-thread-and-order.md:179]: This doesn't clarify the issues in #1167 FTR -- the homeserver still needs to decide on this ordering (and we should define it in the spec). It is fine if this MSC doesn't attempt to do this though.

File: proposals/4033-event-thread-and-order.md [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+### Redacted events
+
+Since we propose that receipts should cache the `order` of their referred-to
+event, this means we do not need to look within a redacted event for its order,
+because the receipt provides it. This avoids needing to preserve the `order`
+property when redacting events.
```
Comment (clokep) [inline proposals/4033-event-thread-and-order.md:185]: I think this is saying that the `order` field should be returned even for redacted events? I don't see why this needs to be called out specifically? (Is `age` not returned for redacted events?)

File: proposals/4033-event-thread-and-order.md [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+Some home servers such as Synapse already have a concept of Stream Order. We
+expect that the order defined here could be implemented using Stream Order.
+
+## Potential issues
+
+This explicitly allows receipts to be inconsistent across federation. In
+practice this is already the case in the wild, and is impossible to solve using
+Stream Order. The problems with using Topological Order (and Sync Order) have
+already been outlined.
```
Comment (clokep) [inline proposals/4033-event-thread-and-order.md:242]: Receipts over federation have some concept of being across multiple events. See how [User Read Receipt](https://spec.matrix.org/v1.7/server-server-api/#definition-mreceipt) has multiple event IDs available to it.

Synapse always sends a single event ID.

File: proposals/4033-event-thread-and-order.md [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+MSC3981 would solve many problems for clients because the events in a thread
+would be returned in Sync Order, but this is not true: the proposal will return
+events in Topological Order, which is useless for determining which events are
+read.
+
+### The server could calculate unread status
+
+We could use the definitions within this proposal but avoid calculating what was
+unread on the client. Instead we could ask the server to figure out which rooms
+are unread.
```
Comment (clokep) [inline proposals/4033-event-thread-and-order.md:264]: Does the server already figure out which rooms are unread? Or are you differentiating between "notification count > 0" and "bold" in Element Web speak?

File: proposals/4033-event-thread-and-order.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+server providing messages in Sync Order, so we could happily fetch messages
+recursively and still be able to slot them into the right thread and ordering.
+
+Note that the expectation (from some client devs e.g. me @andybalaam) was that
+MSC3981 would solve many problems for clients because the events in a thread
+would be returned in Sync Order, but this is not true: the proposal will return
+events in Topological Order, which is useless for determining which events are
+read.
+
+### The server could calculate unread status
```
Comment (clokep) [inline proposals/4033-event-thread-and-order.md:260]: I expected this section to suggest adding a boolean flag to each event's `unsigned` that marked whether the event was before or after the receipt!

File: proposals/4033-event-thread-and-order.md (hunk 3 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: I removed this part of the MSC to simplify it down to just ordering.

File: proposals/4033-event-thread-and-order.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Removed now this MSC is focused on ordering only.

File: proposals/4033-event-thread-and-order.md (hunk 4 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Note added in bbf1c94ab

File: proposals/4033-event-thread-and-order.md (hunk 6 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Removed this sentence.

**Comment by andybalaam** [comment]
> Funnily enough we at Beeper have been using this exact approach (`com.beeper.hs.order` field in the `unsigned` portion) in our Synapse fork and I can say it's been useful for simplifying client development: [beeper/synapse#42](https://github.com/beeper/synapse/pull/42)
> 
> We were also planning on putting an order on read receipts as well, but haven't gotten to it yet.
> 
> Great minds and all that.

That is great news. We can use this as evidence of an implementation, which will help this MSC to progress smoothly.

File: proposals/4033-event-thread-and-order.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+See also [Spec Issue #1167](https://github.com/matrix-org/matrix-spec/issues/1167),
+which calls out this ambiguity about the meaning of "read up to".
+
+### Problems with the current definition
+
+The current definition of *after* is ambiguous, and difficult for clients to
+calculate. It depends on only receiving events via sync, which is impossible
+when we want to do backpagination or to fetch thread messages via the
+`relations` API.
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Yes, clarified in 74933916

File: proposals/4033-event-thread-and-order.md [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+it sends it to any client, and it should never change. It should,
+loosely-speaking, increase for "newer" messages. It may be negative.
+
+The exact meaning of this order is not as important as the fact that it exists
+and is consistent between clients and servers, and events and receipts.
+
+We do not require that order be consistent across federation. The only
+requirement is that all the clients for a user and the one server to which they
+connect agree. For this reason, we propose that `order` be included in an
+event's `unsigned` property.
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Applied in 1b668ce9

File: proposals/4033-event-thread-and-order.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+  },
+  "type": "m.receipt"
+}</pre>
+
+We propose:
+
+* all events should contain an `order` property inside `unsigned`.
+* all receipts should contain an `order` property alongside `ts` inside the
+  information about an event, which is a cache of the `order` property within
+  the referred-to event.
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:149]: Although other users' receipts are less important for the user experience than the current user's, they suffer from the same problems as the current user's, so yes, all receipts.

This means that we always have a consistent view of the read status of other users, but it does mean that we will sometimes see different status from what they themselves see, due to differences across federation. This is currently true, and this MSC does not fix it.

Do you think I should add a comment to this effect somewhere?

File: proposals/4033-event-thread-and-order.md (hunk 8 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:182]: I think this MSC as a whole does clarify #1167 and chooses stream order, but with a caveat that if the answer were topological order, we could live with it. Do you disagree? Do I need to be more explicit?

File: proposals/4033-event-thread-and-order.md (hunk 9 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:188]: No, it is saying that we _don't_ need order to be included with redacted events because if a receipt points at a redacted event, we already have its order within the receipt, so we can still see whether other events are read.

I guess that does leave the question of whether the redacted event itself is read, so it would be convenient if order were included.

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:188]: `age` is returned within a redacted event.

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:188]: The spec does not mention `unsigned` in https://spec.matrix.org/unstable/rooms/v10/#redactions and yet I see it in my redacted events, so I am guessing `unsigned` is considered outside of the purview of the redaction algorithm.

Given that, I will add a note saying that `order` should also be included in redacted events :-)

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Done in d2fe0f41

File: proposals/4033-event-thread-and-order.md (hunk 10 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:247]: I don't understand the relevance of your comment to this MSC. It feels like it could be important, but I'm not getting it, sorry!

File: proposals/4033-event-thread-and-order.md (hunk 11 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:269]: Yes, I am differentiating between notifications and unreads. Currently the server provides no information about unreads.

File: proposals/4033-event-thread-and-order.md (hunk 12 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:265]: Hmm, I guess it could, but I am thinking at the room level.

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:265]: I will clarify that in the title.

**Comment by andybalaam** [comment]
@bradtgmurray is the client code open source?

File: proposals/4033-event-thread-and-order.md (hunk 11 above)
Comment (bradtgmurray) [reply proposals/4033-event-thread-and-order.md:269]: Note https://github.com/matrix-org/matrix-spec-proposals/pull/2654 does exist, but it's not yet merged and may not be.

**Comment by bradtgmurray** [comment]
> @bradtgmurray is the client code open source?

Unfortunately no

File: proposals/4033-event-thread-and-order.md (hunk 10 above)
Comment (clokep) [reply proposals/4033-event-thread-and-order.md:247]: I think I was attempting to say that *some* of the mechanisms exist to have consistent receipts over federation. They're not really implemented / finished, however.

tl;dr I don't think you need to make any changes, but wanted to point it out.

File: proposals/4033-event-thread-and-order.md (hunk 8 above)
Comment (clokep) [reply proposals/4033-event-thread-and-order.md:182]: Maybe use "clarified" instead of "unchanged" (and mark the PR as fixing that issue).

File: proposals/4033-event-thread-and-order.md (hunk 11 above)
Comment (clokep) [reply proposals/4033-event-thread-and-order.md:269]: > Note #2654 does exist, but it's not yet merged and may not be.

I don't think this is actually related -- I think @andybalaam is using "unread" to mean "bold" not having separate unread counts.

File: proposals/4033-event-thread-and-order.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (clokep) [reply proposals/4033-event-thread-and-order.md:149]: I guess I find that surprising due to all the other bits of the MSC talking about your own receipt. I assumed it was a typo.

File: proposals/4033-event-thread-and-order.md [hunk 16]
```diff
@@ -0,0 +1,324 @@
+# MSC4033: Explicit ordering of events for receipts
```
Comment (jplatte) [inline proposals/4033-event-thread-and-order.md:1]: I've heard that Synapse already tracks the timestamp at which is first saw a given event (which is equivalent to `origin_server_ts` if the event came in over client-server API, but not equivalent if it came in over federation).

Wouldn't it make sense to just expose this to clients *as a timestamp* instead of introducing an opaque "order" that's pretty likely to just be that timestamp, but not necessarily across all homeservers so you can't actually rely on it?

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:1]: Some messages reach the homeserver quite late, but are "known" to be "earlier" because they were explicitly requested by this server from another server in order to do backpagination (i.e. fetching older events). In that case, Synapse gives a negative stream order, and this stream order would be better than the timestamp at which Synapse first saw the event.

File: proposals/4033-event-thread-and-order.md (hunk 8 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md]: Clarified in e650a11a and PR marked as fixing #1167

File: proposals/4033-event-thread-and-order.md [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+This explicitly allows receipts to be inconsistent across federation. In
+practice this is already the case in the wild, and is impossible to solve using
+Stream Order. The problems with using Topological Order (and Sync Order) have
+already been outlined.
+
+## Alternatives
+
+### Solves the same problem MSC3981 Relations Recursion tried to solve
+
+This propsal would not replace
```
Comment (ara4n) [inline proposals/4033-event-thread-and-order.md:256]:
```suggestion
This proposal would not replace
```

File: proposals/4033-event-thread-and-order.md [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+order" where late-arriving messages tend to receive higher numbers, or
+"Topological Order" where late-arriving message tend to receive lower numbers.
+
+We believe that it is better to consider late-arriving messages as unread,
+meaning the client has the information that these newly arrived messages have
+not been read and can choose how to display it (or not). This is what leads us
+to suggest Stream Order as the correct choice.
+
+However, if servers choose Topological Order, this proposal still works - we
+just have what the authors consider undesirable behaviour regarding
```
Comment (ara4n) [inline proposals/4033-event-thread-and-order.md:219]: Worth noting that the undesirability of this could be mitigated by using Ranged Read Receipts (currently a protomsc lost at https://hackmd.io/Gxm8zuuSROeencoJ6gjgSg - i.e. you'd get told "you haven't read this range of delayed messages which just turned up yet".

File: proposals/4033-event-thread-and-order.md [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+  }
+}</pre>
+
+Example receipt (changes are highlighted in bold):
+
+<pre>{
+  "content": {
+    "$1435641916114394fHBLK:matrix.org": {
+      "m.read": {
+        "@rikj:jki.re": {
```
Comment (ara4n) [inline proposals/4033-event-thread-and-order.md:139]:
```suggestion
        "@erikj:jki.re": {
```

File: proposals/4033-event-thread-and-order.md [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+    "formatted_body": "&lt;b&gt;This is an example text message&lt;/b&gt;",
+    "msgtype": "m.text"
+  },
+  "event_id": "$143273582443PhrSn:example.org",
+  "origin_server_ts": 1432735824653,
+  "room_id": "!jEsUZKDJdhlrceRyVU:example.org",
+  "sender": "@example:example.org",
+  "unsigned": {
+    "age": 1234,
+    <b>"order": 56764334543</b>
```
Comment (ara4n) [inline proposals/4033-event-thread-and-order.md:108]: This feels like a really nice solution, and I'm pretty confused on how we've got far without having an explicit ordering.  Fully supportive of this :) 🚀

**Review by ara4n — APPROVED** [review]

File: proposals/4033-event-thread-and-order.md (hunk 18 above)
Comment (clokep) [reply proposals/4033-event-thread-and-order.md:219]: @andybalaam and I discussed this a bit, but there's lots of questions around the UX of handling that.

If it is easier to reference by me making a "real" MSC out of that, I can push the current content as a draft PR.

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:219]: I agree that ranged receipts might help with this undesirable behaviour, but I consider this very much a side issue of this MSC, so happy to link to it if you do create it, but I don't see a burning need.

I do note that the current receipts are essentially a special case of ranged receipts, so without having read the proposal, I would expect that the same issues in terms of event order would apply to them, so I'd expect this MSC to become a dependency of that one...

File: proposals/4033-event-thread-and-order.md [hunk 21]
```diff
@@ -0,0 +1,346 @@
...
+Note that the `order` property for a particular event will probably be the same
+for every user, so will be repeated multiple times in an aggregated receipt
+event. This structure was chosen to reduce the chance of breaking existing
+clients by introducing `order` at a higher level.
+
+### Proposed definition of *after*
+
+We propose that the definition of *after* should be:
+
+* Event A is after event B if its order is larger.
```
Comment (bradtgmurray) [inline proposals/4033-event-thread-and-order.md:176]: How should we handle negative orders?

Synapse uses negative orders when it needs to backfill events. One case of this is a server purging history but then backfilling again from federation. When this occurs, these events get negative stream orders, but represent the order in which the backfilled events arrived at the homeserver, with more recent messages getting lower (less legative) values.

This unfortunately means that events with negative stream orders don't really have a relation to their order in the room, and just instead represent the order in which clients attempted to fetch events from a room that was purged. For example, if a client called /context in an arbitrary place in the history and then paginated forwards and backwards we'd get the /context events having the lowest stream ordering followed by the order the client paginated in.

One idea could just be to omit `order` from the events in the cases where the stream order is negative. We can just say that these events are unorderable, and read receipts are best effort at that point. This hopefully isn't too big of a problem, as these events are only likely to occur in rare cases or deep in the rooms history.

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:176]: Good point. I'd be inclined to say that negative order just means "before all receipts" or "always read" so that these messages can never cause a new notification or unread room. Would that make sense?

Comment (bradtgmurray) [reply proposals/4033-event-thread-and-order.md:176]: That makes sense to me, but curious if others weigh in.

Note these events will never come down to clients over sync either or trigger push rules.

Comment (kegsay) [reply proposals/4033-event-thread-and-order.md:176]: So long as the ordering is consistent, I'm happy to have negative values, even if they don't correlate with actual event orderings.

File: proposals/4033-event-thread-and-order.md [hunk 22]
```diff
@@ -0,0 +1,346 @@
+# MSC4033: Explicit ordering of events for receipts
```
Comment (zecakeh) [inline proposals/4033-event-thread-and-order.md:1]: This MSC only mentions receipts, but not fully read markers which are similar but a different section of the spec and have the same issue.

Is it intentionally out of scope of this MSC? If so, it would probably be worth mentioning it.

Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:1]: Good point. I added a note in 605eaddf . I don't want to expand this MSC to have a wider scope, but fixing fully-read markers to have the same `order` information seems like the way to go.

File: proposals/4033-event-thread-and-order.md (hunk 21 above)
Comment (andybalaam) [reply proposals/4033-event-thread-and-order.md:187]: Added some notes about this in d749fb12

File: proposals/4033-event-thread-and-order.md [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+
+The [spec](https://spec.matrix.org/unstable/client-server-api/#receipts) states
+that receipts are "read-up-to" without explaining what order the events are in,
+so it is difficult to decide whether an event is before or after a receipt.
+
+We propose adding an explicit order number to all events, so that it is clear
+which events are read.
```
Comment (timokoesters) [inline proposals/4033-event-thread-and-order.md:8]: The problem with numbers is that you can't fill gaps between them, so if you already have 15 and 16, there's no place inbetween them. Instead, you could treat the order as a string of digits instead with the property that
15 < 151 < 16

For easier understanding, you can compare this to decimal numbers: 0.15 < 0.151 < 0.16

Comment (programmerjake) [reply proposals/4033-event-thread-and-order.md:8]: if you need to be able to insert a new value between any existing pair, you can use [Dyadic Rationals](https://en.wikipedia.org/wiki/Dyadic_rational), basically a value of the form `a * 2^b` where `a` and `b` are integers.

**Comment by bnjbvr** [comment]
As a SDK implementer, I am *super* interested in this proposal, as it could be useful for other purposes too, in addition to making computing the unread badges very simple. For instance, I think this would allow figuring whether we have gaps when reconciling a local/cached timeline of events with the results of a /messages query.

File: proposals/4033-event-thread-and-order.md (hunk 23 above)
Comment (toger5) [reply proposals/4033-event-thread-and-order.md:8]: Using an array as the index is also a concept that can be found in CRDT's.
`[1]`, `[1,1]`, `[1,2]`, `[1,2,1]`,`[2]`

Comment (bnjbvr) [reply proposals/4033-event-thread-and-order.md:8]: In terms of storage/memory used per event (thus bandwidth usage), dyadic rationals sound optimal: sufficiently large space, only require two u32 or two u64. Strings or array representations would require bigger allocations. (Nobody really proposed it, but FP numbers would be a waste in terms of storage space b/o all the `NaN`s + IEEE754 is hard to get right.)

Comment (programmerjake) [reply proposals/4033-event-thread-and-order.md:8]: dyadic rationals only require one byte for the exponent, unless you want to support a value with >256 bits (imo excessive), so they only need 5 bytes (32-bit mantissa) or 9 bytes (64-bit mantissa), though you can definitely use more for alignment or convenience.

File: proposals/4033-event-thread-and-order.md [hunk 24]
```diff
@@ -0,0 +1,386 @@
+# MSC4033: Explicit ordering of events for receipts
```
Comment (erikjohnston) [inline proposals/4033-event-thread-and-order.md:1]: Just as a quick note: I have strong suspicions that: a) this is not possible to do in a way that doesn't limit scalability of servers, b) this isn't necessarily the order in which clients should render events, and c) trying to handle filling in gaps is complicated as you need to deal with edge cases.

I'm sure we can do something close here, but needs investigation from server teams to figure out viability. 

One change that might help a lot here is: instead of inserting an ordering on all events, instead have an opaque "receipt_key" field for events coming down `/sync` or `/messages`, which you can pass to the `/receipt` API. This *is* a lot less powerful than a full linearized history, but feels immediately more viable.

File: proposals/4033-event-thread-and-order.md (hunk 23 above)
Comment (erikjohnston) [reply proposals/4033-event-thread-and-order.md:8]: Note that we will need to be able to handle edge cases where we run out of exponents, e.g. you have an event A and you keep inserting events just after it (which may very well end up a common case). If you only have a one byte exponent then you quickly run out of room.

Comment (programmerjake) [reply proposals/4033-event-thread-and-order.md:8]: if you run out of exponent bits, that also means the mantissa has to be at least 256-bits wide (or something like that), so you will need big-integer arithmetic for that...i think that's likely excessive.

maybe a better idea is to maintain a binary tree with one node per message and have the server send the list of changed nodes (if the message's contents doesn't change, all that needs to be sent is the node ids of the tree's children, since that's all that changes during tree balancing), this allows tree balancing to avoid the tree getting too deep.

messages would then be ordered by their position in an in-order traversal of that tree.

File: proposals/4033-event-thread-and-order.md (hunk 24 above)
Comment (erikjohnston) [reply proposals/4033-event-thread-and-order.md:1]: Oh, also: the server sometimes doesn't know the correct ordering of two historical events immediately (e.g. it has two disconnected chunks of DAG). This is less of a problem for `/sync` and read receipts (as historical events don't go down `/sync` or count as unread, server side at least), but may be more of one for other use cases.

Comment (programmerjake) [reply proposals/4033-event-thread-and-order.md:1]: > Oh, also: the server sometimes doesn't know the correct ordering of two historical events immediately (e.g. it has two disconnected chunks of DAG).

this is easily handled by my [binary tree idea](https://github.com/matrix-org/matrix-spec-proposals/pull/4033#discussion_r1488348049), the server just picks an arbitrary spot in the tree and then changes it when it knows the correct spot. since the tree nodes' order can be changed without changing message/node ids, this makes it more flexible than using a numeric order.

File: proposals/4033-event-thread-and-order.md [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+does not have the relevant event) it should not send the receipt. If a server
+later receives an event, allowing it to provide an order for this receipt, it
+should send the receipt at that time. Rationale: without the order, a receipt is
+not useful to the client since it is not able to use it to determine which
+events are read. If a receipt points at an unknown event, the safest assumption
+is that other events in the room are unread i.e. there is no receipt.
+
+If a receipt is received for an event with negative order, the server should set
+the order in the receipt to zero. All events with negative order are understood
+to be read.
```
Comment (erikjohnston) [inline proposals/4033-event-thread-and-order.md:176]: I think this rapidly runs into issues with threads. While we don't currently actively push for backfilled events, I think we *do* mark threads as unread client side if they have unread events (whether they're backfilled or not). In particular, either a) you incorrectly mark those threads as read, or b) make it impossible for threads to be marked as read (since you can't send a read receipt for the latest event in the thread). This even applies to the main timeline if the latest events in it are only backfilled events.

Separately, while we may be able to fudge something here for read receipts, this does limit the use of the `order` param for other use cases. In particular, it means that you cannot sensibly use the `order` param to order messages within the timeline, as half the messages won't have an ordering at all. As discussed elsewhere, calculating an `order` param for backfilled events is practically impossible without the ability to renumber things later, which adds a lot of complexity.
